provisioning/e2e/_provisioning_e2e.js

```diff
@@ -0,0 +1,145 @@
+      commonName: registrationId,
+      selfSigned: true,
+      days: 10
+    };
+
+    return pem.createCertificateAsync(certOptions)
+      .then(function(result) {
+        return {
+          cert: result.certificate,
+          key: result.clientKey
+        };
+      });
+  };
+
+  var enrollX509 = function(registrationId, deviceId, x509) {
+    var enrollment =  {
+      registrationId: registrationId,
+      deviceId: deviceId,
+      attestation: {
+        type: 'x509',
+        x509: {
+          clientCertificates: {
+            primary: {
+              certificate: x509.cert
+            }
+          }
+        }
+      },
+      initialTwinState: {
+        desiredProperties: {
+          testProp: registrationId + ' ' + deviceId
+        }
+      }
+    };
+
+    return provisioningServiceClient.createOrUpdateIndividualEnrollmentAsync(enrollment);
+  };
+
+  beforeEach (function() {
+    var id = uuid.v4();
+    deviceId = 'deleteme_provisioning_node_e2e_' + id;
+    registrationId = 'reg-' + id;
+  });
+
+  afterEach (function(callback) {
+    debug('deleting enrollment');
+    provisioningServiceClient.deleteIndividualEnrollmentAsync(registrationId)
+      .catch(function() {
+        debug('ignoring deleteIndividualEnrollment error');
+      })
+      .then(function() {
+        debug('deleting device');
+        registry.delete(deviceId);
+      })
+      .catch(function() {
+        debug('ignoring delete error');
+      })
+      .then(function() {
+        debug('done with per-test cleanup');
+        callback();
+      });
+  });
+
+  var assertRegistrationStatus = function(registrationId, expectedStatus, expectedDeviceId) {
+    return provisioningServiceClient.getIndividualEnrollmentAsync(registrationId)
+      .then(function(enrollment) {
+        assert.equal(enrollment.registrationStatus.status, expectedStatus);
+        if (expectedDeviceId) {
+          assert.equal(enrollment.registrationStatus.deviceId, expectedDeviceId);
+        }
+      });
+  };
+
+  provisioningTransports.forEach(function (Transport) {
+    it ('can create an x509 enrollment, register it using ' + Transport.name + ', and verify twin contents', function(callback) {
+      var x509cert;
+
+      debug('creating x509 certificate');
+      createX509Cert(registrationId)
+      .then(function(cert) {
+        x509cert = cert;
+        debug('enrolling');
+        return enrollX509(registrationId, deviceId, cert);
+      })
+      .then(function() {
+        debug('verifying registration status is unassigned');
+        return assertRegistrationStatus(registrationId, 'unassigned');
+      })
+      .then(function() {
+        debug('registering device');
+        var provisioningDeviceClient = Promise.promisifyAll(Transport.createDeviceClient(idScope));
+        return provisioningDeviceClient.registerAsync(registrationId, x509cert, false)
+          .then(function() {
+            return provisioningDeviceClient.endSessionAsync();
```

Comment:
>provisioningDeviceClient.endSessionAsync(); [](start = 19, length = 43)

I'd like it better if we didn't have to force the customer to call this and ended our session cleanly ourselves.

Reply:
OK.  will remove this on the next iteration.

---
In reply to: [148917962](https://github.com/Azure/azure-iot-sdk-node/pull/135#discussion_r148917962) [](ancestors = 148917962)